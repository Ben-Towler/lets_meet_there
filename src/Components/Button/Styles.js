import styled from 'styled-components';

export const StyledButton = styled.button`
  background-color: ${props => props.theme.colors.skyBlue};
  border-radius: ${props => props.theme.borders};
  font-family: ${props => props.theme.fonts[0]};
  font-size: ${props => props.theme.fontSizes.xsmall};
  border: none;
  margin: 20px;
  padding: 0.25em 1em;
`;
