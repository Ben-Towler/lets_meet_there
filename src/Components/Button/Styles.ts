import styled from 'styled-components';

export const StyledButton = styled.button`
  ${props => props.primary && `
    background-color: ${props.theme.colors.skyBlue};
    font-size: ${props.theme.fontSizes.xsmall};
    border: none;
    margin: 20px;
    padding: 0.25em 1em;
  `}
  ${props => props.secondary && `
    background-color: ${props.theme.colors.white};
    border-color: ${props.theme.colors.whiteSmoke};
    font-size: ${props.theme.fontSizes.xsmall};
    margin: 20px;
    padding: 0.25em 1em;
  `}
  ${props => props.tertiary && `
    background-color: ${props.theme.colors.skyBlue};
    font-size: ${props.theme.fontSizes.small};
    border: none;
    height: 46px;
    padding: 1em 2em;
  `}
  border-radius: ${props => props.theme.borders};
  font-family: ${props => props.theme.fonts[0]};
  outline: none;
  color: ${props => props.theme.colors.navyBlue};
  cursor: pointer;
`;
