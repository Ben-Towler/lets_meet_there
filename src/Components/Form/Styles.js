import styled from 'styled-components';
import { PlaneTakeOff } from '@styled-icons/boxicons-solid/PlaneTakeOff';

export const BluePlaneTakeOff = styled(PlaneTakeOff)`
  color: '#5FDAE3';
`;

export const StyledForm = styled.form`

`;

export const StyledLocationSearchWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
`;

export const StyledDateWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-around;
  margin: 1em;
`;

export const Row = styled.div`
  &:not(:first-child) {
    margin-top: 18px;
    &:last-child {
      margin-top: 28px;
    }
  }
`;