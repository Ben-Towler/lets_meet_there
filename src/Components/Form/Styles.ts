import styled from 'styled-components';

export const StyledForm = styled.form`
`;

export const Row = styled.div`
  &:not(:first-child) {
    margin-top: 18px;
    &:last-child {
      margin-top: 28px;
    }
  }
`;