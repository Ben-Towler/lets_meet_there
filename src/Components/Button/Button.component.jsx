import React from "react";
import { StyledButton } from './Styles'

function Button ({children, ...props}) {
  return (
    <StyledButton {...props}>
      {children}
    </StyledButton>
  );
}

export default Button;
