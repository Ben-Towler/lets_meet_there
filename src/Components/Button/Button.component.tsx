import React, {MouseEvent} from 'react';
import { StyledButton } from './Styles';

interface Props {
  children: string;
  type?: 'submit' | 'button' | 'reset';
  onClick?: (event :MouseEvent<HTMLButtonElement>) => void;
  id?: string;
  is?: string;
}

function Button ({children, ...props}: Props): JSX.Element {
  return (
    <StyledButton {...props}>
      {children}
    </StyledButton>
  );
}

export default Button;
