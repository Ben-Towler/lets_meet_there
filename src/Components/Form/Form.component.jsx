import React, { useState } from 'react';
import { StyledForm, } from './Styles';
import { Button } from 'Components';
import moment from 'moment';
import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';
import LocationSearch from '../LocationSearch/locationSearch';

function Form (props) {
  const [focus, setFocus] = useState(null);
  const [formData, setFormData] = useState({});
  const [isFormComplete, setIsFormComplete] = useState(false);


  const [formDates, setFormDates] = useState({
    startDate: null,
    returnDate: null,
  });

  function handleChange (name, googleLocation) {
    let parsedLoc = googleLocation.split(', ');
    const fields = {...formData};
    fields[name] = parsedLoc[0];
    
    setFormData(fields);
    isFormValid(formData)
  }
  
  function isFormValid (fields) {
    for (let i = 0; i < fields.length - 1; i++) {
     if (fields[i].length === 0) {
       setIsFormComplete(false);
     };
    }
    setIsFormComplete(true);
  }

  function handleSubmit (e) {
    e.preventDefault();

    if (isFormComplete) {
      const formCriteria = {
        ...formData,
      }

      if (props.showDateRange) {
        console.log('formDates.startDate', formDates.startDate);
        console.log('formDates.endDate', formDates.endDate);
        formCriteria.startDate = formatDate(formDates.startDate);
        formCriteria.endDate = formatDate(formDates.endDate);
      }

      props.onSubmit(formCriteria);
    }
  }

  function renderFields (fields) {
    return fields.map ( function (field) {
      return (
        <LocationSearch 
          key={field.name}
          field={ field } 
          onChange={handleChange} 
          formData={formData}
          value={formData[field.name]}
        /> 
      )
    })
  }

  function formatDate (date) {
    const formatted = moment(date).format('YYYY MM DD');
    return formatted.split(' ').join('-');
  }

  return (
    <StyledForm data-testid="form" onSubmit={(e) => handleSubmit(e)}>
      {props.hasError ? 'Error' : null}

      {renderFields(props.fields)}

      {props.showDateRange ?
        <DateRangePicker
          startDate={formDates.startDate}
          startDateId="startDate"
          endDate={formDates.endDate} 
          endDateId="endDate"
          onDatesChange={
            function ({ startDate, endDate }) {
              setFormDates({ startDate, endDate })
            }
          }
          showClearDates={true}
          focusedInput={focus}
          onFocusChange={(focus) => {
            setFocus(focus);
          }}
        /> : null
      }

      <Button type="submit">Submit</Button>
    </StyledForm>
  )
}

export default Form;
