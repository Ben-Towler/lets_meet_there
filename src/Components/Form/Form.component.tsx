// @ts-nocheck
import React, { useState } from 'react';
import { StyledForm, } from './Styles';
import { Button } from 'Components';
import moment from 'moment';
import 'react-dates/initialize';
import { DateRangePicker, DateRangePickerShape } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';
import {LocationSearch} from 'Components';
import {FormData} from 'Interfaces/FormData'

export interface Fields {
  key: string;
  title: string;
  name: string;
  type: string;
  placeholder: string;
}

interface Props {
  showDateRange? : boolean;
  onSubmit?: Function
  hasError?: boolean;
  fields?: Fields[];
}

function Form ({showDateRange, onSubmit, hasError, fields}: Props): JSX.Element {
  const [focus, setFocus] = useState(null);
  const [formData, setFormData] = useState<FormData>({location: null, destination: null});
  const [isFormComplete, setIsFormComplete] = useState(false);


  const [formDates, setFormDates] = useState<FormData>({
    startDate: null,
    endDate: null,
  });

  function handleChange (name: string, googleLocation: string) {
    let parsedLoc = googleLocation.split(', ');
    
    const fieldsObj: FormData = {...formData};
    let location = {};
    location[name] = parsedLoc[0];
    
    // console.log("NAME ", name);
    // console.log("LOCATION ", googleLocation);
    console.log('fieldsObj ', fieldsObj);
    console.log('LOCATION ', location);
    setFormData({...fieldsObj, ...location});
    isFormValid(formData)
  }
  
  function isFormValid (fields) {
    console.log('FIELDS', fields);
    for (let i = 0; i < fields.length - 1; i++) {
     if (fields[i].length === 0) {
       setIsFormComplete(false);
     };
    }
    setIsFormComplete(true);
  }

  function handleSubmit (e: Event) {
    
    e.preventDefault();
    
    console.log('HERE');
    if (isFormComplete) {
      console.log('FORM ', isFormComplete);
      const formCriteria: FormData = {
        ...formData,
      }
      
      if (showDateRange) {
        formCriteria.startDate = formatDate(formDates.startDate);
        formCriteria.endDate = formatDate(formDates.endDate);
      }
      console.log('FORM CRITERIA ', formCriteria);
      onSubmit(formCriteria);
    }
  }

  function renderFields (fields: Fields[]) {
    return fields.map ( function (field) {
      return (
        <LocationSearch 
          key={field.name}
          name={field.name}
          field={ field } 
          onChange={handleChange} 
        /> 
      )
    })
  }

  function formatDate (date: string) {
    const formatted = moment(date).format('YYYY MM DD');
    return formatted.split(' ').join('-');
  }

  // interface DRP {
  //   startDate: string,
  //   startDateId: string,
  //   endDate: string,
  //   endDateId: string,
  //   onDatesChange: Function,
  //   showClearDates: Boolean,
  //   focusedInput: [] | null
  //   onFocusChange: Function,
  // }

  interface DRP extends DateRangePickerShape {}
  const DRPProps : DRP = {
    startDate: formDates.startDate, 
    startDateId: 'startDate',
    endDate: formDates.endDate ,
    endDateId: 'endDate',
    onDatesChange: function ({ startDate , endDate }: {startDate: string, endDate: string}) {
        setFormDates({ startDate, endDate })
    },
    showClearDates: true,
    focusedInput: focus,
    onFocusChange: (focus: [] | null) => {
      setFocus(focus);
    }
  }
  return (
    <StyledForm data-testid='form' onSubmit={(e) => handleSubmit(e)}>
      {hasError ? 'Error' : null}

      {renderFields(fields)}

      {showDateRange ?
        <DateRangePicker
          {...DRPProps}
        /> : null
      }

      <Button tertiary={true} type="submit">Submit</Button>
    </StyledForm>
  )
}

export default Form;
