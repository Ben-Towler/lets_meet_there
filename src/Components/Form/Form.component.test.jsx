import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Form from './Form.component';
import LocationSearch from '../LocationSearch/locationSearch';
import Button from '../Button/Button.component';
import { setupGoogleMock } from '../../Mocks/googleMapsApiMock';
import { render } from '@testing-library/react';

configure({adapter: new Adapter()});

beforeAll(() => {
  setupGoogleMock();
});

describe('Component: Form', () => {

  //TODO test the return data from submit

  const formFields = [
    {
      key: 'location',
      title: 'Location',
      name: 'location',
      type: 'text',
      placeholder: 'London...'
    },
    {
      key: 'destination',
      title: 'Destination',
      name: 'destination',
      type: 'text',
      placeholder: 'Barcelona...'
    }
  ];

  const wrapper = shallow(<Form fields={formFields} showDateRange />);

  test('it should render correctly', () => {
    const formTree = renderer
      .create(<Form fields={formFields}/>)
      .toJSON();
    expect(formTree).toMatchSnapshot();
  });

  test('it should render the location search', () => {
    expect(wrapper.containsMatchingElement(<LocationSearch/>)).toEqual(true);
  });

  test('it should render the date range picker', () => {
    const mounted = mount(<Form fields={formFields} showDateRange />)
    expect(mounted.find('DateRangePicker').length).toEqual(1);
  });

  test('it should have a submit button', () => {
    expect(wrapper.containsMatchingElement(<Button>Submit</Button>)).toEqual(true);
  });

  test('it should fire on submit', () => {
    const mockFunction = jest.fn();
    const {getByText, getByTestId} = render (<Form fields={formFields}/>);

    const btn = getByText('Submit');
    const form = getByTestId('form');
    form.onsubmit = mockFunction;
    btn.click();
    expect(mockFunction).toHaveBeenCalled();
  })

});



