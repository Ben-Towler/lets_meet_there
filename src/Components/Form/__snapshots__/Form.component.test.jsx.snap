// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Form it should render correctly 1`] = `
<form
  className="sc-AxjAm fDEmJb"
  onSubmit={[Function]}
>
  <div>
    <div
      className="autocomplete-root"
    >
      <input
        aria-autocomplete="list"
        aria-expanded={false}
        autoComplete="off"
        className="form-control"
        disabled={false}
        name="location"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="London..."
        role="combobox"
        title="Location"
        type="text"
        value=""
      />
      <div
        className="autocomplete-dropdown-container"
      />
    </div>
  </div>
  <div>
    <div
      className="autocomplete-root"
    >
      <input
        aria-autocomplete="list"
        aria-expanded={false}
        autoComplete="off"
        className="form-control"
        disabled={false}
        name="destination"
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Barcelona..."
        role="combobox"
        title="Destination"
        type="text"
        value=""
      />
      <div
        className="autocomplete-dropdown-container"
      />
    </div>
  </div>
  <button
    className="sc-AxgMl ewSVZr"
    type="submit"
  >
    Submit
  </button>
</form>
`;
